### PR TITLE
g.citation: get year from file attributes and enrich bibtex

### LINKED ISSUE #489: [Bug] g.citation: Citation fails for all modules
**g.citation**

**Describe the bug**
g.citation fails for all modules with with the error message:
```The text does not contain date entry```

**To Reproduce**
```
g.extension extension=g.citation
g.citation -as
```

**Expected behavior**
g.citation should be able to provide citation for at least some modules...

**System description (please complete the following information):**
 - Ubuntu 18.04
 - GRASS GIS 7.8.5

### PR DESCRIPTION
**Pull request by ninsbl**
Replaces #509
Fixes #489

### CODE REVIEW

**Comment by neteler** [comment]
I patched `g.citation` locally with this PR but get this output:

```
g.citation -as
@misc{d_barscale,
  title = {{GRASS GIS: d.barscale module}},
  author = {unknown and Hamish Bowman  <!--},
  howpublished = {https://github.com/OSGeo/grass/tree/releasebranch_8_0/display/d.barscale},
  year = {2022}
  note = {Accessed: 2022-01-13},
}
@misc{d_colorlist,
  title = {{GRASS GIS: d.colorlist module}},
  author = {Andreas Lange  <!--},
  howpublished = {https://github.com/OSGeo/grass/tree/releasebranch_8_0/display/d.colorlist},
  year = {2022}
  note = {Accessed: 2022-01-13},
}
@misc{d_colortable,
  title = {{GRASS GIS: d.colortable module}},
  author = {James Westervelt},
  howpublished = {https://github.com/OSGeo/grass/tree/releasebranch_8_0/display/d.colortable},
  year = {2022}
  note = {Accessed: 2022-01-13},
}
...
@misc{r_univar,
  title = {{GRASS GIS: r.univar module}},
  author = {Hamish Bowman and Martin Landa and Ivan Shmakov and Markus Metz  <!--},
  howpublished = {https://github.com/OSGeo/grass/tree/releasebranch_8_0/raster/r.univar},
  year = {2022}
  note = {Accessed: 2022-01-13},
}
@misc{r_unpack,
  title = {{GRASS GIS: r.unpack module}},
  author = {Hamish Bowman and Martin Landa},
  howpublished = {https://github.com/OSGeo/grass/tree/releasebranch_8_0/scripts/r.unpack},
  year = {2022}
  note = {Accessed: 2022-01-13},
}
...
```

So, all dates are the same.

**Comment by ninsbl** [comment]
> So, all dates are the same.

Yes, that is probably because all were installed / build th same day.

If you do:
```
g.extension extension=r.in.nasadem
g.citation r.in.nasadem
```
you should get a differnet date...
Anyway, this issue was reason to work on: https://github.com/OSGeo/grass/pull/2100
The best solution wold be to store the latest change in the manual, and then pick that up in `g.citation`...

**Comment by ninsbl** [comment]
PR updated to reflect changes in manual entries from git if available...

**Comment by ninsbl** [comment]
Maybe tests for e.g. g.citation -a should be added (more that it succeeds than actual values as they may change).

**Comment by ninsbl** [comment]
Now a test suite (and some other minor improvements like e.g. resolving html escape characters in names) have been added.
Should be ready to merge.

Next, the module could be used to unify how author info is given in the manuals...

**Comment by neteler** [comment]
I have tried the testsuite and got:

```
GRASS nc_spm_08_grass7/user1:testsuite > python test_g_citation.py 

[...]

...
F
======================================================================
FAIL: test_addons (__main__.CiteAllCase)
Test that citation information is collected for all core modules
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/gunittest/case.py", line 1403, in assertModule
    module.run()
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/pygrass/modules/interface/module.py", line 823, in run
    self.wait()
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/pygrass/modules/interface/module.py", line 844, in wait
    raise CalledModuleError(
grass.exceptions.CalledModuleError: Module run `g.citation module=g.citation format=bibtex style=harvard1` ended with an error.
The subprocess ended with a non-zero return code: 1. See the following errors:
b'ERROR: Module g.citation: The text does not contain date entry\n'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mneteler/software/grass-addons/src/general/g.citation/testsuite/test_g_citation.py", line 39, in test_addons
    self.assertModule(module)
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/gunittest/case.py", line 1423, in assertModule
    self.fail(self._formatMessage(msg, stdmsg))
AssertionError: Running <g.citation> module ended with non-zero return code (1)
Called (Python): g.citation(module='g.citation', format='bibtex', style='harvard1')
Called (Bash): g.citation module=g.citation format=bibtex style=harvard1
See the following errors:
ERROR: Module g.citation: The text does not contain date entry


======================================================================
FAIL: test_core_modules (__main__.CiteAllCase)
Test that citation information is collected for all core modules
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/gunittest/case.py", line 1403, in assertModule
    module.run()
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/pygrass/modules/interface/module.py", line 823, in run
    self.wait()
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/pygrass/modules/interface/module.py", line 844, in wait
    raise CalledModuleError(
grass.exceptions.CalledModuleError: Module run `g.citation format=bibtex style=harvard1 -a -d` ended with an error.
The subprocess ended with a non-zero return code: 1. See the following errors:
b'ERROR: Module g.citation: The text does not contain date entry\n'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mneteler/software/grass-addons/src/general/g.citation/testsuite/test_g_citation.py", line 24, in test_core_modules
    self.assertModule(module)
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/gunittest/case.py", line 1423, in assertModule
    self.fail(self._formatMessage(msg, stdmsg))
AssertionError: Running <g.citation> module ended with non-zero return code (1)
Called (Python): g.citation(format='bibtex', style='harvard1', flags='ad')
Called (Bash): g.citation format=bibtex style=harvard1 -a -d
See the following errors:
ERROR: Module g.citation: The text does not contain date entry


======================================================================
FAIL: test_core_modules_formats (__main__.CiteAllCase)
Test that citation information is collected for all core modules
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/gunittest/case.py", line 1403, in assertModule
    module.run()
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/pygrass/modules/interface/module.py", line 823, in run
    self.wait()
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/pygrass/modules/interface/module.py", line 844, in wait
    raise CalledModuleError(
grass.exceptions.CalledModuleError: Module run `g.citation format=json style=harvard1 -a -d` ended with an error.
The subprocess ended with a non-zero return code: 1. See the following errors:
b'ERROR: Module g.citation: The text does not contain date entry\n'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mneteler/software/grass-addons/src/general/g.citation/testsuite/test_g_citation.py", line 34, in test_core_modules_formats
    self.assertModule(module)
  File "/home/mneteler/software/grass82/dist.x86_64-pc-linux-gnu/etc/python/grass/gunittest/case.py", line 1423, in assertModule
    self.fail(self._formatMessage(msg, stdmsg))
AssertionError: Running <g.citation> module ended with non-zero return code (1)
Called (Python): g.citation(format='json', style='harvard1', flags='ad')
Called (Bash): g.citation format=json style=harvard1 -a -d
See the following errors:
ERROR: Module g.citation: The text does not contain date entry


----------------------------------------------------------------------
Ran 3 tests in 7.767s
FAILED (failures=3)

```

**Comment by ninsbl** [comment]
Hmm...
Thanks @neteler for testing. Which versions of GRASS GIS and Python do you use on what distro / OS?
On my local Fedora 36 box with Python 3.10 and GRASS GIS 8.2 and 8.3, the tests complete fine.

I see that tests are failing in github CI too (e.g.): https://github.com/OSGeo/grass-addons/actions/runs/3119473894/jobs/5059463116#step:17:40

However, in CI there seems to be a build issue earlier in the process, causing build of manuals to fail:
https://github.com/OSGeo/grass-addons/actions/runs/3119473894/jobs/5059463116#step:13:216
Thus, `g.citation` cannot find an author section in the manual and fails further down the road...

Not sure what causes the build error. Maybe @wenzeslaus has an idea?

**Comment by neteler** [comment]
> Thanks @neteler for testing. Which versions of GRASS GIS and Python do you use on what distro / OS?

I tested with grass82, on Fedora 36 and Python 3.10.7, same as you.

**Comment by ninsbl** [comment]
Hm... It is also still failing in CI. Really weird as I cannot see why tests fail...

**Comment by wenzeslaus** [comment]
My understanding is that the CI is fixed by #806, but you probably already know that.

Check the duplicate commits from the grass8 branch on your branch. I have seen it many times, just never really figured out how people get there :-) I see you merged grass8 into your branch 12 days ago, that was good. Some cherry-pick or rebase later on probably recreated commits from grass8 on your branch. Perhaps doing merge and later rebase is the cause. Any guesses here?

Anyway, I have tested the following on your branch and it gets me to a pretty good situation you can work with (from this [SO answer](https://stackoverflow.com/a/25357146/574907)).

```bash
git reset $(git merge-base grass8 $(git branch --show-current))
```

You will loose all your commit _messages_ and you will have to force push to the remote branch to get back to normal, but otherwise this seems okay.

File: src/general/g.citation/g.citation.py [hunk 1]
```diff
@@ -188,6 +188,35 @@ def clean_line_item(text):
...
+    datetime.datetime(2022, 9, 18, 23, 55, 9)
+    """
+    datetime_capture = (
+        r"^  (Latest change: )(.*)( in commit: ).*"
+        )
+    match = re.search(datetime_capture, text, re.MULTILINE | re.DOTALL | re.IGNORECASE)
+    if match:
+        date_format = "%A %b %d %H:%M:%S %Y"
+    else:
+        datetime_capture = (
```
Comment (tmszi) [inline src/general/g.citation/g.citation.py]: 1. I'm tested this PR with GRASS GIS 8.3.dev version compiled from Git source (last change in module manual is retrieved from commit), **all tests pass** with python3.{8..11}

2. With GRASS GIS 8.2.0 version compiled from tarball without applied this patch https://github.com/OSGeo/grass/releases/download/8.2.0/core_modules_with_last_commit.patch  (last change in module manual is retrieved from the file) **all tests fail** with python3.{8..11}. Same result as @neteler.

Due this line code https://github.com/OSGeo/grass/blob/f9a30a386c46ea1f281a14ad5b9570aa84a078a7/utils/mkhtml.py#L166 which return date time with following format `Wed Oct  5 05:23:22 2022` with 2 white spaces between `Oct\s\s5` (`\s` is white space), regular expression doesn't catch this date time format.

Correct regular expression should be:

`r"^  (Accessed: ).*([A-Z][a-z]{2} [A-Z][a-z]{2}\s\s[0-9]{1,2} [0-9]{2}:[0-9]{2}:[0-9]{2} [0-9]{4}).*"` with two white spaces `\s` between month and day.

Anyway, I suggest that we use the same date time format (`"%A %b %d %H:%M:%S %Y"`) here https://github.com/OSGeo/grass/blob/f9a30a386c46ea1f281a14ad5b9570aa84a078a7/utils/mkhtml.py#L166 as in other cases.

Example:

```python
In [58]: import os

In [59]: from datetime import datetime

In [60]: datetime_format = "%A %b %d %H:%M:%S %Y"

In [61]: datetime.fromtimestamp(os.path.getmtime("/tmp")).strftime(datetime_format)
Out[61]: 'Wednesday Oct 05 06:42:36 2022'
```

**Review by tmszi — CHANGES_REQUESTED** [review]

File: src/general/g.citation/g.citation.py (hunk 1 above)
Comment (ninsbl) [reply src/general/g.citation/g.citation.py]: Thanks, @tmszi  for catching the issue! I tried to address it on the addon side for now, by filtering one or more spaces between month and day ([75f8e72](https://github.com/OSGeo/grass-addons/pull/677/commits/75f8e7253194fd201df30dedcf9f4f3c85fba1ec)). I agree that a consistent format in the manuals would be preferable.
And thanks, @wenzeslaus. It seems I messed up the PR again when I tried to include #806 to get the CI working for this PR... Will try your suggestion to cleanup the mess later today!

**Comment by ninsbl** [comment]
OK, now the single addon module test succeeds: https://github.com/OSGeo/grass-addons/actions/runs/3187995507/jobs/5200154340#step:17:78

However, since all addons are build into the core installation (I guiess), `g.citation -a` fails at least for some addons, like e.g. _i.gabor_:
`ERROR: Module i.gabor: No HTML manual page entry for 'i.gabor'`

I expect it to work with core modules now...

So, I am not sure if this is an issue with the `i.gabor` manual or with `g.citaion` (and the fact that addons are compiled into the core installation, but my hunch is, with investigating, that the issue is with the i.gabor manual...)

**Comment by tmszi** [comment]
> However, since all addons are build into the core installation (I guiess), `g.citation -a` fails at least for some addons, like e.g. _i.gabor_: `ERROR: Module i.gabor: No HTML manual page entry for 'i.gabor'`

Yes you are right. `i.gabor` addon has dependency on the py package `scipy` and inside the CI build env this dependency is missing according https://github.com/OSGeo/grass-addons/blob/b4cfa50ee83494c188c024548a4ff4f0600dc551/.github/workflows/requirements.txt.

File mentioned above should be:

```
matplotlib
numpy
six
scipy
```

What you need is to explore all addons and check their dependencies on py packages, and add the missing ones to the file above. Maybe it would  be good if every addon has function which print list of py packages required for installation via pip as input to a script that will automatically add these dependencies to the file above.

**Comment by tmszi** [comment]
> What you need is to explore all addons and check their dependencies on py packages, and add the missing ones to the file above. Maybe it would be good if every addon has function which print list of py packages required for installation via pip as input to a script that will automatically add these dependencies to the file above.

Better and simple solution should be lazy import additional py packages.  `i.gabor` module has lazy import but with wrong order

https://github.com/OSGeo/grass-addons/blob/b4cfa50ee83494c188c024548a4ff4f0600dc551/src/imagery/i.gabor/i.gabor.py#L198-200

**Comment by tmszi** [comment]
> Ah, OK. Thanks @tmszi ! That is very helpful information. Should we then go on and merge this PR then (you would have to approve as merging is blocked now)?

I had prepared locally PR which standardize date and time format across our all use cases (I'm testing it locally). It is change in GRASS GIS `mkhtml.py` file (one line changes in date time format). My opinion is that, we can incorporate this changes into this PR  (we will adjust regular expression to catch this standardized date and time format). What is the your opinion?

> I assume that there are some other addons with similar (and other issues) in the manuals. Probably more appropriate to fix those in another PR?

Yes I agree.

**Comment by tmszi** [comment]
Could you try this with your branch https://github.com/OSGeo/grass-addons/pull/677#issuecomment-1267802225, please?

**Comment by ninsbl** [comment]
Hm... I did not actually mean to merge and close the PR. That happen after doing:

```
git reset $(git merge-base grass8 $(git branch --show-current))
git push --force
```

**Comment by wenzeslaus** [comment]
I'm sorry for confusion. That is expected behavior. You need to commit the changes between reset and push. `git status` is your friend after every Git command.

But never mind, new PR is a good opportunity to start a fresh discussion.

**Pull request closed without merging**